examples/dnn_dcgan_train_ex.cpp

```diff
@@ -0,0 +1,272 @@
+    "Unsupervised Representation Learning with Deep Convolutional Generative Adversarial Networks"
+    by Alec Radford, Luke Metz, Soumith Chintala.
+
+    The main idea is that there are two neural networks training at the same time:
+    - the generator is in charge of generating images that look as close as possible as the
+      ones from the dataset.
+    - the discriminator will decide whether an image is fake (created by the generator) or real
+      (selected from the dataset).
+
+    Each training iteration alternates between training the discriminator and the generator.
+    We first train the discriminator with real and fake images and then use the gradient from
+    the discriminator to update the generator.
+
+    In this example, we are going to learn how to generate digits from the MNIST dataset, but
+    the same code can be run using the Fashion MNIST datset:
+    https://github.com/zalandoresearch/fashion-mnist
+*/
+
+#include <algorithm>
+#include <iostream>
+
+#include <dlib/data_io.h>
+#include <dlib/dnn.h>
+#include <dlib/gui_widgets.h>
+#include <dlib/matrix.h>
+
+using namespace std;
+using namespace dlib;
+
+// We start by defining a simple visitor to disable bias learning in a network.  By default,
+// biases are initialized to 0, so setting the multipliers to 0 disables bias learning.
+class visitor_no_bias
+{
+public:
+    template <typename input_layer_type>
+    void operator()(size_t , input_layer_type& ) const
+    {
+        // ignore other layers
+    }
+
+    template <typename T, typename U, typename E>
+    void operator()(size_t , add_layer<T, U, E>& l) const
+    {
+        set_bias_learning_rate_multiplier(l.layer_details(), 0);
+        set_bias_weight_decay_multiplier(l.layer_details(), 0);
+    }
+};
+
+// Some helper definitions for the noise generation
+const size_t noise_size = 100;
+using noise_t = std::array<matrix<float, 1, 1>, noise_size>;
+
+noise_t make_noise(dlib::rand& rnd)
+{
+    noise_t noise;
+    for (auto& n : noise)
+    {
+        n = rnd.get_random_gaussian();
+    }
+    return noise;
+}
+
+// A convolution with custom padding
+template<long num_filters, long kernel_size, int stride, int padding, typename SUBNET>
+using conp = add_layer<con_<num_filters, kernel_size, kernel_size, stride, stride, padding, padding>, SUBNET>;
+
+// A transposed convolution to with custom padding
+template<long num_filters, long kernel_size, int stride, int padding, typename SUBNET>
+using contp = add_layer<cont_<num_filters, kernel_size, kernel_size, stride, stride, padding, padding>, SUBNET>;
+
+// The generator is made of a bunch of deconvolutional layers.  Its input is a 1 x 1 x k noise
+// tensor, and the output is the generated image.  The loss layer does not matter for the
+// training, we just stack a compatible one on top to be able to have a () operator on the
+// generator.
+using generator_type =
+    loss_binary_log_per_pixel<
+    htan<contp<1, 4, 2, 1,
```

Comment:
Instead of `htan` you could use `sig`, which makes outputs appear between 0 and 1.  Then the code to convert to images is just `255 * out`, which is simpler.  This isn't going to make training converge faster, just make the code more straightforward.

Reply:
Indeed, way simpler :)